dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorPower.java

```diff
@@ -285,4 +299,336 @@ public List<Event> visitDat3mCAS(Dat3mCAS e) {
+        Store store = newRMWStoreExclusive(address, dummyReg, null, true);
+        Label label = newLabel("FakeDep");
+        Event fakeCtrlDep = newFakeCtrlDep(resultRegister, label);
+
+        Fence optionalMemoryBarrierBefore = mo.equals(Tag.Linux.MO_MB) ? Power.newSyncBarrier()
+                : mo.equals(Tag.Linux.MO_RELEASE) ? Power.newLwSyncBarrier() : null;
+        Fence optionalMemoryBarrierAfter = mo.equals(Tag.Linux.MO_MB) ? Power.newSyncBarrier()
+        		: mo.equals(Tag.Linux.MO_ACQUIRE)? Power.newISyncBarrier() : null;
+
+        
+        return eventSequence(
+                optionalMemoryBarrierBefore,
+                load,
+                localOp,
+                store,
+                fakeCtrlDep,
+                label,
+                optionalMemoryBarrierAfter
+        );
+	};
+
+	@Override
+	public List<Event> visitRMWOpReturn(RMWOpReturn e) {
+		Register resultRegister = e.getResultRegister();
+		IOpBin op = e.getOp();
+		IExpr value = (IExpr) e.getMemValue();
+		IExpr address = e.getAddress();
+		String mo = e.getMo();
+		
+        Register dummyReg = e.getThread().newRegister(resultRegister.getPrecision());
+        Local localOp = newLocal(resultRegister, new IExprBin(dummyReg, op, value));
+
+        // Power does not have mo tags, thus we use null
+        Load load = newRMWLoadExclusive(dummyReg, address, null);
+        Store store = newRMWStoreExclusive(address, resultRegister, null, true);
+        Label label = newLabel("FakeDep");
+        Event fakeCtrlDep = newFakeCtrlDep(dummyReg, label);
+
+        Fence optionalMemoryBarrierBefore = mo.equals(Tag.Linux.MO_MB) ? Power.newSyncBarrier()
+                : mo.equals(Tag.Linux.MO_RELEASE) ? Power.newLwSyncBarrier() : null;
+        Fence optionalMemoryBarrierAfter = mo.equals(Tag.Linux.MO_MB) ? Power.newSyncBarrier()
+        		: mo.equals(Tag.Linux.MO_ACQUIRE)? Power.newISyncBarrier() : null;
+
+        
+        return eventSequence(
+                optionalMemoryBarrierBefore,
+                load,
+                localOp,
+                store,
+                fakeCtrlDep,
+                label,
+                optionalMemoryBarrierAfter
+        );
+	};
+
+	@Override
+	public List<Event> visitRMWFetchOp(RMWFetchOp e) {
+		Register resultRegister = e.getResultRegister();
+		IExpr value = (IExpr) e.getMemValue();
+		IExpr address = e.getAddress();
+		String mo = e.getMo();
+		
+        Register dummy = resultRegister;
+		if(resultRegister == value){
+            dummy = e.getThread().newRegister(resultRegister.getPrecision());
+        }
+
+		Load load = newRMWLoadExclusive(dummy, address, null);
+        Local optionalUpdateReg = dummy != resultRegister ? newLocal(resultRegister, dummy) : null;
+        Store store = newRMWStoreExclusive(address, new IExprBin(dummy, e.getOp(), value), null, true);
+        Label label = newLabel("FakeDep");
+        Event fakeCtrlDep = newFakeCtrlDep(dummy, label);
```

Comment:
I get that there can be issues when the written `value` expression coincides with the `resultRegister`.
Can a similiar issue happen when `address` coincides with `resultRegister` (does that even make sense?).
And why do the other `RMW` events do not take the possibility of `resultRegister == value` into account?

Lastly, I noticed that the naming `e.getMemValue()` is very misleading as it does not correspond to the value the operation will eventually write to memory! It might be better to rename this to just `e.getValue()` or `e.getOperandValue()`, but this is an issue orthogonal to this PR.

Reply:
This is the same pattern that is being used in `visitorLKMM`. My understanding is that the other methods do not need to perform the check because they do not fetch the old value.

Not comparing with `address` can indeed generate problems (good catch!), and this is already a bug in `visitorLKMM`. The following program fails under LKMM when it should clearly be safe.
```
C C-atomic-fetch-simple-01

{
  atomic_t x[2] = {0,0};
}

P0(atomic_t *x) {
  int r0; int r1;
  r0 = 1;
  r0 = atomic_fetch_add(2,x+r0);
}

exists (0:r0 != 0 \/ x[1] != 2)
```
I also think it is not enough to compare `resultRegister == value`. We should probably use `value.getRegs().contains(resultRegister)` which is the one I used to solve the problem with `address` from above.

Yes, the `e.getMemValue()` is an unfortunately artifact of inheriting from MemEvent.

Reply:
I think the other events do use `value` in some way after performing the load into `resultRegister`. That means the load can affect the `value`, which should never be intended.
Maybe we shouldn't bother with the check and always create dummy registers to just avoid any of this special reasoning? The only disadvantage should be that we end up with a lot of dummy registers. But then I'm wondering if we could even reuse the same dummy register (per thread) over and over again (this might negatively affect static analyses though).

Also, I noticed that `RMWAbstract` has its own member `value` but it overrides `getMemValue` to return it. I think it would be reasonable to just add a new getter and setter instead of reusing `getMemValue` there.